### PR TITLE
enh(Change Logs): Insert change logs in /configuration/hosts endpoints

### PR DESCRIPTION
**Pull request by dmyios**
## Description

Added change log insertion in 
- `configuration/hosts POST`
- `configuration/hosts/{host_id} PATCH & DELETE` endpoints

**Fixes** # MON-107626

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [x] master

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Review by a-launois — APPROVED** [review]

File: centreon/src/Core/ActionLog/Infrastructure/Repository/DbWriteActionLogRepository.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        } catch (\Throwable $ex) {
+            $this->error($ex->getMessage(), ['trace' => $ex->getTraceAsString()]);
+
+            if (! $aleadyInTransction) {
+                $this->db->rollBack();
+            }
+
+            throw $ex;
+        }
```
Comment (Drago37) [inline centreon/src/Core/ActionLog/Infrastructure/Repository/DbWriteActionLogRepository.php:132]: vraiment dommage de ne pas capter la véritable exception et de ne pas renvoyer une exception, à la lecture ce serait lus clair, et nous pourrions mieux gérer les erreurs : )

**Review by jeremyjaouen — APPROVED** [review]

File: centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function add(NewHost $host): int
+    {
+        try {
+            $hostId = $this->writeHostRepository->add($host);
+            if ($hostId === 0) {
+                throw new RepositoryException('Host ID cannot be 0');
```
Comment (Drago37) [inline centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php:74]: ah bah voilà ! du coup pourquoi ne pas l'utiliser dans DbWriteActionLogRepository ? 😄

File: centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+            $diff = array_diff_assoc($updatedHostDetails, $currentHostDetails);
+
+            $this->writeHostRepository->update($host);
+
+            if (array_key_exists('isActivated', $diff)) {
+                if ((bool) $diff['isActivated']) {
+                    $actionType = ActionLog::ACTION_TYPE_ENABLE;
+                } else {
+                    $actionType = ActionLog::ACTION_TYPE_DISABLE;
+                }
```
Comment (Drago37) [inline centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php:157]: tu pourrais faire une ternaire ici pour simplifier ton code :)

File: centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                    if (is_string($value[0])) {
+                        $value = '!' . implode(
+                            '!',
+                            str_replace(["\n", "\t", "\r"], ['#BR#', '#T#', '#R#'], $value)
+                        );
+                    }
+
+                    if ($value[0] instanceof HostEvent) {
+                        $value = HostEventConverter::toString($value);
+                    }
```
Comment (Drago37) [inline centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php:266]: enchainé avce des elseif plutôt non ? ca évite des sous if type Hadoken (si t'as la réf) 😃

File: centreon/src/Core/ActionLog/Domain/Model/ActionLog.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+{
+    public const ACTION_TYPE_ADD = 'a';
+    public const ACTION_TYPE_CHANGE = 'c';
+    public const ACTION_TYPE_DELETE = 'd';
+    public const ACTION_TYPE_ENABLE = 'enable';
+    public const ACTION_TYPE_DISABLE = 'disable';
+
+    private ?int $id;
+
+    private \DateTime $creationDate;
```
[the remark this answers is not in the document]
Comment (dmyios) [reply centreon/src/Core/ActionLog/Domain/Model/ActionLog.php:35]: generally we don't do phpdoc for typed properties

File: centreon/src/Core/ActionLog/Application/Repository/WriteActionLogRepositoryInterface.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * For more information : contact@centreon.com
+ *
+ */
+declare(strict_types=1);
+
+namespace Core\ActionLog\Application\Repository;
+
+use Core\ActionLog\Domain\Model\ActionLog;
+
+interface WriteActionLogRepositoryInterface
```
[the remark this answers is not in the document]
Comment (dmyios) [reply centreon/src/Core/ActionLog/Application/Repository/WriteActionLogRepositoryInterface.php:27]: We do CQRS pattern and split write and read repositories

File: centreon/src/Core/ActionLog/Application/Repository/WriteActionLogRepositoryInterface.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /**
+     * @param ActionLog $actionLog
+     *
+     * @throws \Throwable
```
[the remark this answers is not in the document]
Comment (dmyios) [reply centreon/src/Core/ActionLog/Application/Repository/WriteActionLogRepositoryInterface.php:32]: `\Throwable` is a "wider" exception it allows to catch all sorts of Exceptions (`PDOException`, `RepositoryException`, etc.)

File: centreon/src/Core/ActionLog/Infrastructure/Repository/DbWriteActionLogRepository.php [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        return (int) $this->db->lastInsertId();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function addActionDetails(ActionLog $actionLog, array $details): void
+    {
+        if ($details === []) {
+            return;
```
[the remark this answers is not in the document]
Comment (dmyios) [reply centreon/src/Core/ActionLog/Infrastructure/Repository/DbWriteActionLogRepository.php:88]: In this case we do early return, if `$details` array is empty we simply do not proceed with DB insertion and return.

File: centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php (hunk 4 above)
Comment (dmyios) [reply centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostLogRepository.php:266]: True, I will try to refacto in my next PRs

**Pull request merged**